build/test/generate/expected_outputs_test.dart

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:build/build.dart';
+import 'package:build_test/build_test.dart';
+import 'package:test/test.dart';
+
+void main() {
+  test('replaces file extensions', () {
+    _expectOutputs(
+      {
+        '.dart': ['.g.dart']
+      },
+      _asset('lib/foo.dart'),
+      [_asset('lib/foo.g.dart')],
+    );
+  });
+
+  group('capture groups', () {
+    test('can match files in directory', () {
+      _expectOutputs(
+        {
+          'proto/{{}}.proto': ['lib/src/generated/{{}}.dart']
+        },
+        _asset('proto/service.proto'),
+        [_asset('lib/src/generated/service.dart')],
+      );
+    });
+
+    test('can match files in subdirectories', () {
+      _expectOutputs(
+        {
+          'proto/{{}}.proto': ['lib/src/generated/{{}}.dart']
+        },
+        _asset('proto/services/auth.proto'),
+        [_asset('lib/src/generated/services/auth.dart')],
+      );
+    });
+
+    test('must have unique names', () {
+      expect(
+        () => expectedOutputs(
+            TestBuilder(buildExtensions: {'{{}}/{{}}/foo.txt': []}),
+            _asset('foo.txt')),
+        throwsArgumentError,
+      );
+
+      expect(
+        () => expectedOutputs(
+            TestBuilder(buildExtensions: {'{{x}}/{{x}}/foo.txt': []}),
+            _asset('foo.txt')),
+        throwsArgumentError,
+      );
+    });
+
+    test('need to be used in outputs', () {
+      expect(
+        () => expectedOutputs(
+          TestBuilder(buildExtensions: {
+            '{{}}.txt': ['invalid.txt']
+          }),
+          _asset('foo.txt'),
+        ),
+        throwsArgumentError,
+      );
+
+      expect(
+        () => expectedOutputs(
+          TestBuilder(buildExtensions: {
+            '{{x}}/{{y}}.txt': ['{{x}}.txt']
+          }),
+          _asset('foo.txt'),
+        ),
+        throwsArgumentError,
+      );
+    });
+
+    test('may not be used in the same output multiple times', () {
+      expect(
+        () => expectedOutputs(
+          TestBuilder(buildExtensions: {
+            '{{}}.txt': ['{{}}/{{}}/.foo']
+          }),
+          _asset('foo.txt'),
+        ),
+        throwsArgumentError,
+      );
+    });
+
+    test('output may not use capture groups not used in the input', () {
+      expect(
+        () => expectedOutputs(
+          TestBuilder(buildExtensions: {
+            '{{input}}.txt': ['{{output}}.foo']
```

Comment:
```suggestion
            '.txt': ['{{output}}.foo']
```

(this test wasn't currently distinguishing between two potential errors)

Reply:
Good catch, this check was also missing in the code because we just use raw suffixes if there's no capture group in the input.